### ISSUE TITLE
Unable to retrieve data from experiment

### ISSUE THREAD

**Issue by nclaidiere** [open-cogsci/omm-server#150]
Hi,
Since yesterday we are unable to retrieve data from an ongoing experiment. When we try to download data we get the following error message:
![data download problem](https://user-images.githubusercontent.com/50901388/235079956-2d924c41-90dd-412f-a0d3-480d83e61730.PNG)

Computer ressources are fine, I updated the computer and docker with no effect. The size of the DB is 0.5GB so not that big.

When I look at the log from mysql on docker, it seems the DB is shutting down when I press download, this is what happens:

2023-04-28 09:25:32 2023-04-28T07:25:32.408203Z 0 [System] [MY-013172] [Server] Received SHUTDOWN from user <via user signal>. Shutting down mysqld (Version: 8.0.30).
2023-04-28 09:25:34 2023-04-28T07:25:34.409559Z 0 [Warning] [MY-010909] [Server] /usr/sbin/mysqld: Forcing close of thread 10  user: 'omm'.
2023-04-28 09:25:34 2023-04-28T07:25:34.409749Z 0 [Warning] [MY-010909] [Server] /usr/sbin/mysqld: Forcing close of thread 11  user: 'omm'.
2023-04-28 09:25:36 2023-04-28T07:25:36.896947Z 0 [System] [MY-010910] [Server] /usr/sbin/mysqld: Shutdown complete (mysqld 8.0.30)  MySQL Community Server - GPL.
2023-04-28 09:25:40 2023-04-28T07:25:40.586218Z 0 [Warning] [MY-011068] [Server] The syntax '--skip-host-cache' is deprecated and will be removed in a future release. Please use SET GLOBAL host_cache_size=0 instead.
2023-04-28 09:25:40 2023-04-28T07:25:40.587234Z 0 [System] [MY-010116] [Server] /usr/sbin/mysqld (mysqld 8.0.30) starting as process 1
2023-04-28 09:25:40 2023-04-28T07:25:40.614388Z 1 [System] [MY-013576] [InnoDB] InnoDB initialization has started.
2023-04-28 09:25:40 2023-04-28 07:25:40+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 8.0.30-1.el8 started.
2023-04-28 09:25:40 2023-04-28 07:25:40+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
2023-04-28 09:25:40 2023-04-28 07:25:40+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 8.0.30-1.el8 started.
2023-04-28 09:25:40 '/var/lib/mysql/mysql.sock' -> '/var/run/mysqld/mysqld.sock'
2023-04-28 09:25:45 2023-04-28T07:25:45.702853Z 1 [System] [MY-013577] [InnoDB] InnoDB initialization has ended.
2023-04-28 09:25:49 2023-04-28T07:25:49.304692Z 0 [Warning] [MY-010068] [Server] CA certificate ca.pem is self signed.
2023-04-28 09:25:49 2023-04-28T07:25:49.304747Z 0 [System] [MY-013602] [Server] Channel mysql_main configured to support TLS. Encrypted connections are now supported for this channel.
2023-04-28 09:25:49 2023-04-28T07:25:49.445530Z 0 [Warning] [MY-011810] [Server] Insecure configuration for --pid-file: Location '/var/run/mysqld' in the path is accessible to all OS users. Consider choosing a different directory.
2023-04-28 09:25:49 2023-04-28T07:25:49.662301Z 0 [System] [MY-011323] [Server] X Plugin ready for connections. Bind-address: '::' port: 33060, socket: /var/run/mysqld/mysqlx.sock
2023-04-28 09:25:49 2023-04-28T07:25:49.662442Z 0 [System] [MY-010931] [Server] /usr/sbin/mysqld: ready for connections. Version: '8.0.30'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  MySQL Community Server - GPL.

This is the list of 'vulnerabilities':
![image](https://user-images.githubusercontent.com/50901388/235083907-92e6accc-5cd4-46e8-b8fa-e0f33bdb8fd7.png)

It seems the experiment is working fine as long as I don't try to download the data.

**Comment by smathot** [comment]
For now, restart the server and don't try to download anything. That way you at least won't lose any data while we're resolving this. @dschreij Can you take a look at this asap? The culprit is almost certainly this line:

> 2023-04-28 09:25:32 2023-04-28T07:25:32.408203Z 0 [System] [MY-013172] [Server] Received SHUTDOWN from user .

But where does the SHUTDOWN signal come from?

**Comment by dschreij** [comment]
My question exactly. I have searched for other reports of this error but haven't found anything yet in this context. A first hunch is that the database is overloaded by the download operation (composing the data file is expensive), but I don't expect MySQL to shut down like this. Do you also have more log lines from before:

> 2023-04-28 09:25:32 2023-04-28T07:25:32.408203Z 0 [System] [MY-013172] [Server] Received SHUTDOWN from user . Shutting down mysqld (Version: 8.0.30).

I might shed some light on what is causing the shutdown.

I intend to continue exploring this problem tomorrow evening. I will see if I can dig up the VPN credentials, I was sent a long time ago, but I am hesitant to try to reproduce this on the production server at the risk of data loss as Sebastiaan mentioned.

**Comment by dschreij** [comment]
I did some more searching, and I am not thinking of issues caused by too limited memory. Linux (which Docker uses behind the scenes) has something called the OOM killer, which terminates processes if it runs out of memory. Since this happens when data is being downloaded for this particular experiment, and not on other occasions (right?), I have the following questions to see if it is worth following this assumption:

- Is the collected data set of this experiment exceptionally large (as in, much bigger than other experiments you have worked with)
- I know I may have asked this before, but just to be sure (and have this information also attached to this issue): what are the specs of the computer that is running the server / performing the data collection (RAM, CPU, etc.)

See also:
https://serverfault.com/questions/564378/mysql-shut-down-it-self-periodically
https://lwn.net/Articles/317814/

I will see if I can find more.

**Comment by nclaidiere** [comment]
Hi,

So I have stopped trying to do anything, in case I get things worse. The experiment seems to be going on fine but we don't know if the data is being saved properly and we can't follow the progress of the experiment. So Daniel, is there a way to access the DB without using OMM? In particular, would it be possible to save a copy? I tried using workbench a few days ago but couldn't get it to connect and then I stopped doing anything. If we can access the DB through any other mean we can always fix the problem later.

To answer your questions.

Here is the computer configuration, as I said it's not overloaded:
Nom de l'appareil	CCDP-51
Processeur	Intel(R) Core(TM) i7-4790 CPU @ 3.60GHz   3.60 GHz
Mémoire RAM installée	16,0 Go
ID de périphérique	0DB9487E-1B69-4168-87B7-11B093507FED
ID de produit	00330-50000-00000-AAOEM
Type du système	Système d’exploitation 64 bits, processeur x64
Stylet et fonction tactile	Prise en charge du stylet et de la fonction tactile avec 10 points de contact
Édition	Windows 10 Professionnel
Version	21H2
Installé le	‎25/‎03/‎2021
Build du système d’exploitation	19044.2846
Expérience	Windows Feature Experience Pack 120.2212.4190.0

Docker seems to have enough space:
![image](https://user-images.githubusercontent.com/50901388/235340869-e9f182da-a1c1-40ae-9465-a03c90a4dccc.png)

But there is a lot of critical issues:
![image](https://user-images.githubusercontent.com/50901388/235340958-c9064663-6a81-4423-bc2f-cf6c2ce092b4.png)

All these packages need updating? or not?

Clicking everywhere I just found this:
![image](https://user-images.githubusercontent.com/50901388/235341103-483d8220-2396-4366-8696-0095752967c3.png)

This is much higher... if the mysql data is 4GB that's a lot.

Regarding the experiment itself, it's not extremely high, we are logging ~50 variables.

Regarding the log, I have attached a copy of the file. 
[log omm server.txt](https://github.com/open-cogsci/omm-server/files/11360797/log.omm.server.txt)

One significant piece of info is this I think:
2023-04-28 11:10:35 [Error: ENOENT: no such file or directory, unlink '/img/public/files/28/data_20230427091220.csv'] {
2023-04-28 11:10:35   errno: -2,
2023-04-28 11:10:35   code: 'ENOENT',
2023-04-28 11:10:35   syscall: 'unlink',
2023-04-28 11:10:35   path: '/img/public/files/28/data_20230427091220.csv',
2023-04-28 11:10:35   status: 500
2023-04-28 11:10:35 }
This is what appears on the OMM server interface...

Let me know if I can do anything else.

Thanks,
nico

**Comment by dschreij** [comment]
Hi Nico,

> So Daniel, is there a way to access the DB without using OMM? In particular, would it be possible to save a copy?

Yes, it should be. If I check the `docker-compose.yml`, the Mysql database is exposed on port 3306. I just tried it myself and could connect to the database with the following settings:

![image](https://user-images.githubusercontent.com/689044/235374518-48fca7d0-f309-4af3-8ae4-0e61f72f7845.png)

The database won't show up Workbench as a result of its scan (probably because it is running in Docker). You can then export all data from the menu Server > Data Export

![image](https://user-images.githubusercontent.com/689044/235374573-aeac1468-941b-42cf-861c-9c39ea3beb25.png)

> Here is the computer configuration, as I said it's not overloaded:

I believe you when you say this, but one thing I did notice is

> Mémoire RAM installée 16,0 Go

16GB should be enough for this server, however, Docker (or rather WSL2 on which it runs) is only allowed to use a limited proportion of the total system memory. According to [the docs](https://learn.microsoft.com/en-us/windows/wsl/wsl-config#configure-global-options-with-wslconfig) no more than 50% or 8 GB of RAM (whichever is less).

![image](https://user-images.githubusercontent.com/689044/235375074-41a4d39b-e0c2-4a9a-9596-ddc0679579e3.png)

You can check the usage by opening Task manager and checking processes. If you sort by memory, you will see something like Vmmem or VmmemWSL at the top:

![image](https://user-images.githubusercontent.com/689044/235374731-7feb5885-27f3-4b94-90ee-b32933ea64d8.png)

So even though you have 16GB, Mysql may only have 8 GB at its disposal, which is a bit cramped for a DB with the size of 4 GB.
One thing we can try to test this memory hypothesis is to increase the memory docker is allowed to use. If you go inside docker and click the cog/settings icon, you will see a menu item Resources. You can either adjust the settings directly there if your Docker is *not* running in Windows Subsystem for Linux (WSL2), or if it is (in the case of my screenshot), you need to change the configuration in a file

![image](https://user-images.githubusercontent.com/689044/235375006-ea5ffe5e-0757-4b9d-a70e-20a5a3ac73ec.png)

If you see the same contents as in my screenshot, you will need to create or edit your .wslconfig file in `C:\Users\<UserName>\.wslconfig` and add a section in which you specify the allowed memory that can be used, e.g.

```
[wsl2]

memory=12G
```

(See https://learn.microsoft.com/en-us/windows/wsl/wsl-config#example-wslconfig-file for a full example)

If you then restart both WSL and your docker containers, it should have more available memory. If this was the problem, then the download should proceed now, if not, then it has a different cause and we should hunt for other leads.

Just to shortly touch up onto your other points:

> But there is a lot of critical issues:

These are vulnerability scans which indicate if other open source packages used have security issues. This is less scary than it looks and often remediated by updating dependency packages, but I doubt it is the cause of the current problem.

> One significant piece of info is this I think:
> 2023-04-28 11:10:35 [Error: ENOENT: no such file or directory, unlink '/img/public/files/28/data_20230427091220.csv'] {
> 2023-04-28 11:10:35 errno: -2,
> 2023-04-28 11:10:35 code: 'ENOENT',
> 2023-04-28 11:10:35 syscall: 'unlink',
> 2023-04-28 11:10:35 path: '/img/public/files/28/data_20230427091220.csv',
> 2023-04-28 11:10:35 status: 500
> 2023-04-28 11:10:35 }
> This is what appears on the OMM server interface...

This is probably a result of my code not catching an error correctly and I will fix that. However, I suspect that it is the result of the problem at hand rather than the cause of it, but I will check this of course.

I hope this story is clear. Let me know if I can help you with these or the next steps.

**Comment by nclaidiere** [comment]
Hi Daniel, thanks for all this.

Regarding memory, Vmem is more than 3GB, and I can imagine this goes up when I press the download data button:
![image](https://user-images.githubusercontent.com/50901388/235418822-28b4ea5d-d7a6-41c7-91d6-6dd0abe1ced0.png)

I changed wsl2 mem for 12GB, restarted the computer, same result. But indeed the memory used goes up to 7GB when I try to download the data, so I guess if the 12GB setting is not taken into account that could still be the issue. Last time, when docker was using all the RAM I never managed to set the wsl mem, it never took the values into account... so I don't know. Is there a way to check the memory limit from docker? 

Regarding the workbench connection, I copied your settings, retrieved the omm password from docker, and tested the connection, it failed:
![image](https://user-images.githubusercontent.com/50901388/235422570-cf8f1f3f-d977-4f74-bb4c-f21fe0bf4f3e.png)

Here is the docker compose file:
[docker-compose.txt](https://github.com/open-cogsci/omm-server/files/11364115/docker-compose.txt)

I compared to the one on github and the ports are missing:
    ports:
      - 3306:3306
So that must be the reason no? I updated the docker compose file, but how can I make it take it into account? I am afraid of ruining the DB...

Thanks

**Comment by dschreij** [comment]
Hi Nico,

Alright, at least we now know limited memory isn't the (complete) issue.

>I compared to the one on github and the ports are missing:
>ports:
> - 3306:3306
> So that must be the reason no? I updated the docker compose file, but how can I make it take it into account? I am afraid of > ruining the DB...

Yes that's it. if it just said 3306 then the port wasn't exposed to programs outside of docker, but if you enter the address this way you should be able to access the database from outside docker. You can safely execute `docker compose restart` (see [docs](https://docs.docker.com/engine/reference/commandline/compose_restart/)) to restart all containers. All the data is stored in separate volume containers, so they are pretty safe from whatever happens to the main app containers. Otherwise, you wouldn't be able to perform upgrades either.

**Comment by dschreij** [comment]
Hi Nico,

Alright, at least we now know limited memory isn't the (complete) issue.

>I compared to the one on github and the ports are missing:
>ports:
> - 3306:3306
> So that must be the reason no? I updated the docker compose file, but how can I make it take it into account? I am afraid of > ruining the DB...

Yes that's it. if it just said 3306 then the port wasn't exposed to programs outside of docker, but if you enter the address this way you should be able to access the database from outside docker. You can safely execute `docker compose restart` (see [docs](https://docs.docker.com/engine/reference/commandline/compose_restart/)) to restart all containers. All the data is stored in separate volume containers, so they are pretty safe from whatever happens to the main app containers. Otherwise,you wouldn't be able to perform upgrades either.

**Comment by smathot** [comment]
It seems clear that the underlying issue is that the system is running out of memory when the download query is executed. The only real way to solve this is by upgrading the memory in the system. So @nclaidiere, here's what I would do right now:

1. Create a dump of the database using the instructions that @dschreij provided above. (The fact that this is necessary implies that right now there is no backup system in place (!), which is something to get on top of asap once this particular issue is resolved.)
2. Upgrade the system memory to (at least) 64 Gb. In most cases, you can simply buy a memory card, open up the computer, and insert the memory card into a slot.

Once this is done, the question is whether memory consumption will continue to increase as the database continues to grow. Downloading larger data sets will require more memory—that is normal. However, downloading a data set of a fixed size should require an approximately fixed amount of memory, regardless of how much *other* data there is in the database. Do you see the distinction here?

To test this, I would occasionally download the data from an experiment that is finished (so that you're always downloading the same amount of data) and note down how much memory consumption increases when you do this. If memory consumption increases over time (i.e. as the amount of *other* data in the database increases), then we need to take a look at that because this will cause issues in the long term. But if it doesn't, then we don't.

**Comment by dschreij** [comment]
Like Sebastiaan said, upgrading the RAM of the system and drastically upping the amount Docker is allowed to use can alleviate this issue. Memory has never been as cheap as it currently is (you can get 32GB of RAM for around 50 euros) so that's good news.

Another option is to not run MySQL in Docker but on Windows natively. Then it has access to full system memory, but getting MySQL to run under Windows and configuring and maintaining it will be added overhead that you must consider. Even in that case, 16 GB of RAM may not be enough at some point in the near future, if your dataset keeps growing. Regardless it's a good idea to increase the memory of the system to _at least_ 32 GB of RAM.

**Comment by nclaidiere** [comment]
Hi, ok for all of the above. So first a copy of the DB.

When I run docker compose restart I get:
PS C:\omm-server-master> docker compose restart
[+] Running 0/1
 - Container omm-server-master-mailhog-1  Restarting                                                                                                                                                                                    1.3s
Error response from daemon: Cannot restart container a22d241f9040c105c4458b94f2c5e00f73004c418241b620e530e56d6112e13e: driver failed programming external connectivity on endpoint omm-server-master-mailhog-1 (e5c272020e94369800606d66692424931dd01096ff88394ae1d9dcaa81eb92c7): Bind for 0.0.0.0:3001 failed: port is already allocated

And it says in the doc:
If you make changes to your compose.yml configuration, these changes are not reflected after running this command. 

Is that the right way of doing it then?

**Comment by dschreij** [comment]
Right, the error you are seeing only concerns mailhog, which is currently not important.

You can also try:
```
docker compose stop mysql
docker compose start mysql
```
and if that doesn't work
```
docker compose down mysql
docker compose up -d mysql
```
Or you can even do this from the Docker GUI, I suppose, by pressing the `stop` button next to the mysql container (which is equal to docker compose stop).

**Comment by nclaidiere** [comment]
Hi Daniel,

I have tried everything... every time using localhost, the ip address 192.168.201.51

This:
docker compose stop mysql
docker compose start mysql
has no effect. I have also tried shutting docker down and the computer down and restarting.

And I get this with the other commands:
PS C:\omm-server-master> docker compose down mysql
unknown command "mysql" for "docker compose down"
PS C:\omm-server-master> docker compose up -d mysql
[+] Running 0/0
 - Container mysql  Creating                                                                                       0.0s
Error response from daemon: Conflict. The container name "/mysql" is already in use by container "17ce64d56255a10b2ca42eb3f526c57380260c83e83b0155f1fcd6ebfd7d19de". You have to remove (or rename) that container to be able to reuse that name.

One command doesn't exist, the other returns the same error message as before.

With docker shutdown:
PS C:\omm-server-master> docker compose up -d mysql
error during connect: this error may indicate that the docker daemon is not running: Get "http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.24/containers/json?all=1&filters=%7B%22label%22%3A%7B%22com.docker.compose.config-hash%22%3Atrue%2C%22com.docker.compose.project%3Domm-server-master%22%3Atrue%7D%7D": open //./pipe/docker_engine: The system cannot find the file specified.

When I do this, this seems to create another instance of mysql????

![image](https://user-images.githubusercontent.com/50901388/235732315-155d24b2-f541-46de-93fa-f6fa38227b85.png)

**Comment by nclaidiere** [comment]
Hi, this is getting weirder...

![image](https://user-images.githubusercontent.com/50901388/235849294-37971c67-220d-48c5-91ca-b6ede0e38d48.png)

It's clear I don't understand what I am doing and I am going to mess things up. So I am going to stop doing anything and try to put a new memory slot in tomorrow, hope for the best. 

Daniel, if you understand what's going on please let me know.

Cheers

**Comment by dschreij** [comment]
Hi Nico,

Your screenshot is showing the images tab, so that means images that have been built/created/downloaded but are not running. You can safely delete the topmost images that are not in use. If you view the Containers tab, you see all the images that are currently running (in a container).

I looked up these commands in the documentation (see also [this](https://stackoverflow.com/questions/42149529/how-to-reload-environment-variables-in-docker-compose-container-with-minimum-dow)) and found out you can simply reload the configuration by executing `docker compose up -d <service_name>` again. I just tried it myself and it does the trick. So, if you change something to the `mysql` section of `docker-compose.yml` and execute `docker compose up -d mysql`, the container should be recreated with the new variables.

If it has worked, you will see the ports listed in the Containers tab of the docker GUI:

![image](https://user-images.githubusercontent.com/689044/235850555-a13a0abd-4759-4214-af43-6738b92141e9.png)

**Comment by dschreij** [comment]
> Daniel, if you understand what's going on please let me know.

Docker groups all containers under the name of the folder in which the `docker-compose.yml` file is found, so can it be that you copied this to another location and executed a docker compose up from there? Nothing to be alarmed of at least. If you just execute docker-compose up from you omm-server-master folder, all containers should be grouped under there again. All data should safely be stored in a volume container and will still be accessible once mysql and omm server containers are running again.

**Comment by nclaidiere** [comment]
Hi,

I increased the RAM of the computer from 16 to 24 GB with no effect... I am starting to think this might not be linked to the RAM no? Also, everything I have tried to access the DB using workbench has failed.

I am pretty desperate now. Is there a way to transfer/save docker? would that be an option?

**Comment by dschreij** [comment]
Relates to #92

**Comment by dschreij** [comment]
@nclaidiere I just pushed a fix for the error mentioning a datafile cannot be deleted. Can you try it out and tell me if it works now? I couldn't try it myself as it's hard to reproduce the error.
I also updated the results view to use the more optimized query.

**Comment by smathot** [comment]
Thanks for getting on top of this, guys. Does this mean that the source of the problem was a bug in the server after all, rather than a memory issue?

**Comment by nclaidiere** [comment]
Hi, First, thanks again to Daniel for taking the time to look this up directly and solve it, I really appreciate it. If I understand correctly, the 'out of sort memory' in https://github.com/open-cogsci/omm-server/issues/92 was the culprit. The lack of memory prevented a file from being created, after that, the software simply stopped because it was trying to delete a non-existent file, every time creating a log file that was 1GB (the size of the DB). 
The 'out of sort' memory issue is not resolved, we upped the size of the sort memory in the docker-compose file a lot but we will reach that ceiling again (@dschreij correct me if I am wrong). This is a known issue.
@dschreij fixed the 'delete file that doesn't exist' issue with the server.
@dschreij  also improved the DB view that retrieves the data from the DB.

**Comment by dschreij** [comment]
> Does this mean that the source of the problem was a bug in the server after all, rather than a memory issue?

There were two separate problems here. First the database had too little memory to perform its operation which caused it to crash (or shutdown). My guess is also that this shutdown was graceful, hence no error was reported to the server. This proceeded as usual, but then bumped into the situation that a file wasn't there which it did expect. So, another defensive programming lesson for me was to place the delete operation in a try-catch block as well.

> The 'out of sort' memory issue is not resolved, we upped the size of the sort memory in the docker-compose file a lot but we will reach that ceiling again (@dschreij correct me if I am wrong). This is a known issue.

Well, sort of. The sort memory was by default 128kb, and we increased it to (the community recommended) 5MB, so we should be good for a while. So we may hit the ceiling again at some point, but I guess it will take a while. There were some other memory settings in the mysql configuration that were conservative and we also increased those. Finally, there was a typo in the .wslconfig (First line contained `[wsl12]` instead of `[wsl2]`) so I don't think it ever was processed correctly. All in all, the database should have much more breathing space now.

The memory settings have been incorporated in the updated `docker-compose.yml` file.

**Comment by smathot** [comment]
Ok, that's clear. 👍 

Just to remind ourselves: the out-of-sort memory occurs, as far as we could tell, only for experiments in which the same jobs are repeated over and over, which is the case in a typical 'infinite' background task where jobs are reset each time. @nclaidiere Was this the case for the experiment that triggered the issue now?

Avoiding this implementation by having multiple identical jobs, such that individual jobs are not repeated too often, will also effectively avoid the out-of-sort memory. I'm not sure this is necessary now with the optimizations and increased memory limit, but do keep this in mind in case the problem resurfaces.

**Comment by nclaidiere** [comment]
"Was this the case for the experiment that triggered the issue now?"
No, but we repeated the same joblist 80 times, that isn't huge but come to think of it that's probably one of the biggest we've ran so far (apart from the background tasks).

**Comment by nclaidiere** [comment]
Hi, This is also in relation to issue #152.

I am trying to download data from an experiment that has been running for awhile (it's infinitely looping) and when I try, there is no "out of sort memory" anymore but the csv file is never generated. Vmmem took 25% of CPU and 10GB of RAM for 2.5h and never managed to output the data, then I stopped it. So I think we've reached the limit of the system.

Notably, this does seem to be independent of the 1GB mysql redolog files. I purged all the log files and pressing the download button does not generate a new 1GB log file, whereas they were generated when the experiment was running (it doesn't run anymore). So these seems to be two separate issues.
